File: foodify_api/src/models/Order.model.ts
import mongoose from "mongoose";

const orderItemSchema = new mongoose.Schema({
    foodId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: "Food",
        required: true
    },
    name: { type: String, required: true },
    price: { type: Number, required: true },
    quantity: { type: Number, required: true },
    image: { type: String, required: true },
});

const orderSchema = new mongoose.Schema({
    userId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: "User",
        required: true
    },
    restaurantId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: "Restaurant",
        required: true
    },
    restaurantName: { type: String, required: true },
    items: [orderItemSchema],
    subtotal: { type: Number, required: true },
    deliveryFee: { type: Number, required: true },
    totalAmount: { type: Number, required: true },
    deliveryAddress: { type: String, required: true },
    phone: { type: String, required: true },
    paymentMethod: {
        type: String,
        enum: ["Cash on Delivery", "eSewa", "Khalti"],
        default: "Cash on Delivery"
    },
    status: {
        type: String,
        enum: ["pending", "preparing", "out_for_delivery", "delivered", "cancelled"],
        default: "pending"
    },
    orderDate: { type: Date, default: Date.now },
    deliveryDate: { type: Date },
}, { timestamps: true });

export const OrderModel = mongoose.model("Order", orderSchema);

File: foodify_mobile/lib/features/orders/data/models/order_model.dart
import '../../domain/entities/order_entity.dart';

class OrderItemModel extends OrderItemEntity {
  const OrderItemModel({
    required super.foodId,
    required super.name,
    required super.price,
    required super.quantity,
    required super.image,
  });

  factory OrderItemModel.fromJson(Map<String, dynamic> json) {
    return OrderItemModel(
      foodId: json['foodId'] ?? '',
      name: json['name'] ?? '',
      price: (json['price'] ?? 0).toDouble(),
      quantity: json['quantity'] ?? 0,
      image: json['image'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'foodId': foodId,
      'name': name,
      'price': price,
      'quantity': quantity,
      'image': image,
    };
  }
}

class OrderModel extends OrderEntity {
  const OrderModel({
    required super.id,
    required super.userId,
    required super.restaurantId,
    required super.restaurantName,
    required super.items,
    required super.subtotal,
    required super.deliveryFee,
    required super.totalAmount,
    required super.deliveryAddress,
    required super.phone,
    required super.paymentMethod,
    required super.status,
    required super.orderDate,
    super.deliveryDate,
  });

  factory OrderModel.fromJson(Map<String, dynamic> json) {
    // Handle restaurantId - can be String or Object
    String restaurantId = '';
    if (json['restaurantId'] is String) {
      restaurantId = json['restaurantId'];
    } else if (json['restaurantId'] is Map) {
      restaurantId = json['restaurantId']['_id'] ?? '';
    }

    return OrderModel(
      id: json['_id'] ?? '',
      userId: json['userId'] ?? '',
      restaurantId: restaurantId, // ✅ FIXED
      restaurantName: json['restaurantName'] ?? '',
      items:
          (json['items'] as List<dynamic>?)
              ?.map((item) => OrderItemModel.fromJson(item))
              .toList() ??
          [],
      subtotal: (json['subtotal'] ?? 0).toDouble(),
      deliveryFee: (json['deliveryFee'] ?? 0).toDouble(),
      totalAmount: (json['totalAmount'] ?? 0).toDouble(),
      deliveryAddress: json['deliveryAddress'] ?? '',
      phone: json['phone'] ?? '',
      paymentMethod: json['paymentMethod'] ?? 'Cash on Delivery',
      status: json['status'] ?? 'pending',
      orderDate: json['orderDate'] != null
          ? DateTime.parse(json['orderDate'])
          : DateTime.now(),
      deliveryDate: json['deliveryDate'] != null
          ? DateTime.parse(json['deliveryDate'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'userId': userId,
      'restaurantId': restaurantId,
      'restaurantName': restaurantName,
      'items': items.map((item) => (item as OrderItemModel).toJson()).toList(),
      'subtotal': subtotal,
      'deliveryFee': deliveryFee,
      'totalAmount': totalAmount,
      'deliveryAddress': deliveryAddress,
      'phone': phone,
      'paymentMethod': paymentMethod,
      'status': status,
      'orderDate': orderDate.toIso8601String(),
      if (deliveryDate != null) 'deliveryDate': deliveryDate!.toIso8601String(),
    };
  }
}
